File: back/src/main/java/com/back/domain/match/matchRequest/service/MatchRequestRetryProcessor.java
package com.back.domain.match.matchRequest.service;

import com.back.domain.match.matchRequest.entity.MatchRequest;
import com.back.domain.match.matchRequest.repository.MatchRequestRepository;
import com.back.global.exception.ServiceException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.ApplicationContext;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

// MatchRequestRetryProcessor.java
@Slf4j
@Component
@RequiredArgsConstructor
public class MatchRequestRetryProcessor {
    private final MatchRequestRepository matchRequestRepository;
    private final ApplicationContext applicationContext;

    // 매칭 신청 트랜잭션 커밋(afterCommit) 흐름에서 호출되므로, 요청 스레드를 블로킹하지 않도록 비동기로 실행한다.
    @Async
    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void retryOne(Long matchRequestId) {
        MatchRequest matchRequest = matchRequestRepository.findByIdWithMember(matchRequestId)
                .orElseThrow(() -> new ServiceException("404-1", "매칭 요청을 찾을 수 없습니다."));
        applicationContext.getBean(MatchRequestService.class).tryMatch(matchRequest);
    }
}

File: back/src/main/kotlin/com/back/domain/match/matchRequest/service/MatchRequestRetryProcessor.kt
package com.back.domain.match.matchRequest.service

import com.back.domain.member.member.entity.Industry
import org.springframework.context.ApplicationContext
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Component
import java.util.UUID

@Component
class MatchRequestRetryProcessor(
    private val applicationContext: ApplicationContext
) {
    // 요청 스레드를 블로킹하지 않도록 비동기 실행. @Transactional은 걸지 않는다 -
    // Redisson 락 대기 중에는 커넥션을 쥐지 않아야 하기 때문.
    @Async
    fun retryOne(matchRequestId: UUID, industry: Industry) {
        applicationContext.getBean(MatchRequestService::class.java).tryMatch(matchRequestId, industry)
    }
}
